reactor-core/src/test/java/reactor/core/publisher/FluxRefCountGraceTest.java

```diff
@@ -35,6 +35,153 @@
+					//here we use replay
+					.replay()
+					.refCount(1, Duration.ofSeconds(200));
+
+			return Flux.merge(
+					multiplex.take(1),
+					Mono.delay(Duration.ofSeconds(5)).flatMapMany(ignore -> multiplex
+							.materialize().elapsed().map(Object::toString))
+			);
+		})
+		            .expectNext("hello")
+		            .thenAwait(Duration.ofSeconds(5))
+		            //here we expect an immediate replay of the hello + error
+		            .expectNext("[0,onNext(hello)]")
+		            .expectNext("[0,onError(java.lang.RuntimeException: boom)]")
+		            .expectComplete()
+		            .verify(Duration.ofMillis(500));
+
+		assertThat(sourceSubscriptions).as("source subscriptions").hasValue(1);
+	}
+
+	@Test
+	void completeInterruptsGracePeriodPublish() {
+		AtomicInteger sourceSubscriptions = new AtomicInteger();
+		StepVerifier.withVirtualTime(() -> {
+			Flux<Object> multiplex = Flux
+					.concat(
+							Flux.just("hello"),
+							Mono.delay(Duration.ofSeconds(3)).then()
+					)
+					.doOnSubscribe(s -> sourceSubscriptions.incrementAndGet())
+					//here we use publish
+					.publish()
+					.refCount(1, Duration.ofSeconds(200));
+
+			return Flux.merge(
+					multiplex.take(1),
+					Mono.delay(Duration.ofSeconds(5))
+					    .flatMapMany(ignore -> multiplex
+					    .materialize().elapsed().map(Object::toString))
+			);
+				})
+		            .expectNext("hello")
+		            .thenAwait(Duration.ofSeconds(5))
+		            //here we expect that there is a reconnection, so the hello is followed by a 3s pause and onComplete
+		            .expectNext("[0,onNext(hello)]")
+		            .expectNoEvent(Duration.ofSeconds(3))
+		            .expectNext("[3000,onComplete()]")
+		            .expectComplete()
+		            .verify(Duration.ofMillis(500));
+
+		assertThat(sourceSubscriptions).as("source subscriptions").hasValue(2);
+	}
+
+	@Test
+	void completeInterruptsGracePeriodReplay() {
+		AtomicInteger sourceSubscriptions = new AtomicInteger();
+		StepVerifier.withVirtualTime(() -> {
+			Flux<Object> multiplex = Flux
+					.concat(
+							Flux.just("hello"),
+							Mono.delay(Duration.ofSeconds(3)).then()
+					)
+					.doOnSubscribe(s -> sourceSubscriptions.incrementAndGet())
+					//here we use replay
+					.replay()
+					.refCount(1, Duration.ofSeconds(200));
+
+			return Flux.merge(
+					multiplex.take(1),
+					Mono.delay(Duration.ofSeconds(5)).flatMapMany(ignore -> multiplex
+							.materialize().elapsed().map(Object::toString))
+			);
+		})
+		            .expectNext("hello")
+		            .thenAwait(Duration.ofSeconds(5))
+		            //here we expect an immediate replay of the hello + complete
+		            .expectNext("[0,onNext(hello)]")
+		            .expectNext("[0,onComplete()]")
+		            .expectComplete()
+		            .verify(Duration.ofMillis(500));
+
+		assertThat(sourceSubscriptions).as("source subscriptions").hasValue(1);
+	}
+
+	@Test
+	void shouldReconnect() {
```

Comment:
if the above tests satisfy you @osi I'll probably remove this one

Reply:
They look good. Your tests are more comprehensive than what I had put together to demonstrate the problem (distilled from a larger body of work).